dev/AppNotifications/AppNotificationUtility.cpp

```diff
@@ -237,14 +240,101 @@ std::wstring Microsoft::Windows::AppNotifications::Helpers::GetDisplayNameBasedO
+// we should just load that image file size and ignore DPI.
+inline HICON LoadBetterIconFromResource(_In_ HICON hIcon, int cx, int cy, UINT uFlags /*= 0*/)
 {
-    // THROW_HR_IF_MSG(E_UNEXPECTED, VerifyIconFileExtension(iconFilePath));
+    HICON hIconNew = nullptr;
 
-    return E_NOTIMPL;
+    ICONINFOEX iconInfoEx = {};
+    iconInfoEx.cbSize = sizeof(iconInfoEx);
+    if (GetIconInfoEx(hIcon, &iconInfoEx))
+    {
+        if (S_OK != SHDefExtractIcon(iconInfoEx.szModName, -iconInfoEx.wResID, uFlags, &hIconNew, nullptr, MAKELONG(cx, cy)))
+        {
+            DestroyIcon(hIconNew);
+            hIconNew = nullptr;
+        }
+
+        THROW_IF_WIN32_BOOL_FALSE(DeleteObject(iconInfoEx.hbmColor));
+        THROW_IF_WIN32_BOOL_FALSE(DeleteObject(iconInfoEx.hbmMask));
+    }
+    return hIconNew;
 }
 
+inline HICON RetrieveIcon()
+{
+    std::wstring processPath{};
+    THROW_IF_FAILED(wil::GetModuleFileNameExW(GetCurrentProcess(), nullptr, processPath));
+
+    HICON hIcons[1];
+    ExtractIconExW(processPath.c_str(), 0, hIcons, nullptr, 1);
+
+    int iconSizes{ 0 };
+    HICON hIconNew{ nullptr };
+    int const rgSnapSize[] = { 256, 196, 128, 64, 48, 32, 24, 16 };
+    do
+    {
+        // Load a better image, looking for an exact match above 256x256, or something that is close once we get below 48px (unscaled).
+        hIconNew = LoadBetterIconFromResource(hIcons[0], rgSnapSize[iconSizes], rgSnapSize[iconSizes], (rgSnapSize[iconSizes] > 48) ? GIL_EXACTSIZEONLY : 0);
+        iconSizes++;
+    } while ((hIconNew == nullptr) && (iconSizes < ARRAYSIZE(rgSnapSize)));
+
+    if (hIconNew != nullptr)
+    {
+        // Clear the NIIF_USER flag since we need to now destroy the newly loaded icon.
+        //WI_ClearFlag(dwInfoFlags, NIIF_USER);
+        hIcons[0] = hIconNew;
+    }
+
+    return hIcons[0];
+}
+
+inline std::wstring RetrieveLocalFolderPath()
+{
+    wil::unique_cotaskmem_string localAppDataPath;
+    THROW_IF_FAILED(SHGetKnownFolderPath(FOLDERID_LocalAppData, 0, NULL, &localAppDataPath));
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft
+    std::path localFolderPath{ std::wstring(localAppDataPath.get()) + Microsoft::Windows::AppNotifications::Helpers::c_localMicrosoftFolder };
+    THROW_HR_IF(ERROR_FILE_NOT_FOUND, !std::exists(localFolderPath));
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft\WindowsAppSDK
+    localFolderPath.append(LR"(WindowsAppSDK)");
+    if (!std::exists(localFolderPath))
+    {
+        std::create_directory(localFolderPath);
+    }
+
+    return std::wstring(localFolderPath.c_str());
+}
+
+HRESULT Microsoft::Windows::AppNotifications::Helpers::RetrieveAssetsFromProcess(_Out_ Microsoft::Windows::AppNotifications::Helpers::AppNotificationAssets& assets) noexcept try
+{
+    wil::unique_hicon hIcon{ RetrieveIcon() };
+    THROW_HR_IF_MSG(E_UNEXPECTED, hIcon == nullptr, "Icon not found");
```

Comment:
At this layer we shouldn't propagate messages to the developer. It can be confusing reading "Icon not found" but Register() succeeded (and app notifications still show up an icon).

Reply:
makes sense